keps/sig-apps/20190410-configmap-secret-orchestration.md

```diff
@@ -0,0 +1,214 @@
+
+We would like to solve these issues by adding support to the orchestration resources to support orchestration support around config/secrets.
+
+
+## Motivation
+
+Currently there are a mishmash of workarounds to the orchestration objects not supporting orchestration of configmaps/secrets.
+
+Several different solutions or workarounds out of tree are being used. See Alternatives section for details. None of these are widely deployed. Since most workloads need this kind of orchestration, having it as a base Kubernetes functionality would be a significant benefit to users.
+
+### Goals
+
+* Enhance the deployment/statefulset/daemonset with a flag asking for the workload to be automatically updated when a specified configmap/secret changes.
+* Enhance the deployment/statefulset/daemonset with a flag asking for the specified configmap/secret's current state to be snapshotted and follow the orchestrated workload (roll-forward/roll-back)
+* Implementations of the api changes
+
+### Non-Goals
+
+* Out of tree solutions. For applications to be portable, it really needs to be in Kubernetes proper so application developers know the functionality is always there.
+
+## Proposal
+
+The ConfigMapVolumeSource and SecretVolumeSource objects will get two new, optional fields:
+ * snapshot: boolean
+ * watch: boolean
+
+Both will default to false for compatibility.
+
+snapshot can be either true or false. Watch can only be true if snapshot is also true. This ensures immutability.
+
+The fields will be ignored by all objects other then Deployment, DaemonSet and StatefulSet.
+
+DaemonSet/StatefulSet controllers will be modified such that:
+ * During a "change", on seeing a snapshot=true flag on a configmap/secret, a copy of the configmap/secret will be made with a unique name. This unique name will be stored in the corresponding ControllerRevision.
+ * All pods created will reference the unique configmap/secret snapshot name, not the base name.
+ * When a ControllerRevision is deleted, a deletion of the corresponding configmap/secret snapshots will also be issued.
+ * When an object with a watch flag of true is created/updated, watches on the specified configmap/secret will be added. Any watch triggered will be treated as a "change" of the object.
```

Comment:
watch seems redundant to me, the new controller that fulfills this proposal should be able to watch changes to configmaps/secrets that need to be snapshot.